### PR TITLE
[FIX] website_slides_survey: compute left attempts correctly

### PR DESCRIPTION
**Pull request by Mtaylorr**
Issue:
======
- When a user attempt a certification it will count as 1st attempt and
then do retry directly it will count as second attempt : fine.
- When a user attempt a certification it will count as 1st attempt and
  then exit the page and go to the certification again it will count as
  1st attempt too. So the user can have unlimited number of attempts
  like this.

Steps to reproduce the issue:
=============================
- Install survey , elearning
- Go to a course with a certification and do the certification, exit the
  page and enter to it again.
- Go to the user page and see the certifications you will see they both
  counted as 1st attempt.

Expected behavior:
==================
- when a user leaves a course (willingly or not) we remove the
  relationship between the user_input and slide_partner_id.
- To get the remaining attempts we just get the user_input_ids of the
  slide_partner because the old ones are removed.

opw-3498054

### CODE REVIEW

**Review by flch-odoo — COMMENTED** [review]
Good work, @Mtaylorr , thank you!

**Comment by Mtaylorr** [comment]
@flch-odoo thanks for the review , I fixed all the comments.

File: addons/website_slides_survey/models/slide_channel.py [hunk 1]
```diff
@@ -2,9 +2,30 @@
...
     _inherit = 'slide.channel'
 
     nbr_certification = fields.Integer("Number of Certifications", compute='_compute_slides_statistics', store=True)
+
+    def _remove_membership(self, partner_ids):
+        """Clear certifications' user_inputs invite_token.
+        
+        Deleting the user_inputs' invite_token allows to keep track of the current
+        pool of attempts allowed since the user (last) joined the course, as only 
+        those will have an invite token, that will  be reused until there are no
```
Comment (flch-odoo) [inline addons/website_slides_survey/models/slide_channel.py:18]: As you'll fix for the runbot, there is also this one ;)
```suggestion
        those will have an invite token, that will be reused until there are no 
```

**Comment by Mtaylorr** [comment]
@flch-odoo  I fixed the style issues.

**Comment by nle-odoo** [comment]
seems ready to merge

robodoo delegate+

File: addons/website_slides_survey/models/slide_channel.py [hunk 2]
```diff
@@ -2,9 +2,29 @@
...
+        pool of attempts allowed since the user (last) joined the course, as only
+        those will have a slide_id."""
+        removed_channel_partner_domain = []
+        for channel in self:
+            removed_channel_partner_domain = expression.OR([
+                removed_channel_partner_domain,
+                [('partner_id', 'in', partner_ids),
+                 ('channel_id', '=', channel.id)]
+            ])
+        if (removed_channel_partner_domain):
```
Comment (flch-odoo) [inline addons/website_slides_survey/models/slide_channel.py:26]:
```suggestion
        if removed_channel_partner_domain:
```

File: addons/website_slides_survey/models/slide_channel.py [hunk 3]
```diff
@@ -2,9 +2,29 @@
...
+            slide_partner_ids = self.env['slide.slide.partner'].sudo().search(
```
Comment (flch-odoo) [inline addons/website_slides_survey/models/slide_channel.py:27]: It is short-lived but let it be obvious if someone works on this later.
```suggestion
            slide_partner_ids_sudo = self.env['slide.slide.partner'].sudo().search(
```

File: addons/website_slides_survey/tests/test_course_certification_failure.py [hunk 4]
```diff
@@ -98,11 +101,14 @@ def test_course_certification_failure_flow(self):
 
         self.slide_certification.with_user(self.user_portal)._generate_certification_url()
         self.assertTrue(channel_partner.active, 'Portal user membership should be a unarchived upon joining the course once again')
-        self.assertEqual(1, len(slide_partner.user_input_ids.filtered(lambda user_input: user_input.state != 'done')), 'A new user input should have been automatically created upon slide view')
+        self.assertEqual(1, len(slide_partner.user_input_ids.filtered(lambda user_input: user_input.slide_id)), 'A new user input should have been automatically created upon slide view')
 
         # Step 8: fill in the created user_input with correct answers this time
-        self.fill_in_answer(slide_partner.user_input_ids.filtered(lambda user_input: user_input.state != 'done'), certification.question_ids, good_answers=True)
+        self.fill_in_answer(slide_partner.user_input_ids.filtered(
+            lambda user_input: user_input.slide_id), certification.question_ids, good_answers=True)
```
Comment (flch-odoo) [inline addons/website_slides_survey/tests/test_course_certification_failure.py:108]: Unless it is a more recent feature, this is shorter:
 ```suggestion
        self.fill_in_answer(slide_partner.user_input_ids.filtered('slide_id'), certification.question_ids, good_answers=True)
```

**Review by flch-odoo — COMMENTED** [review]
Thank you for the update, @Mtaylorr !

I have a few mainly style-related suggestions but apart from that, it's what we discussed.
Please also update the title of the commit as we do not impact `survey` here.

@awa-odoo, could you please confirm that this is an acceptable fix for the problem we discussed last week?

File: addons/website_slides_survey/controllers/survey.py [hunk 5]
```diff
@@ -7,8 +7,8 @@
 class Survey(main.Survey):
     def _prepare_survey_finished_values(self, survey, answer, token=False):
         result = super(Survey, self)._prepare_survey_finished_values(survey, answer, token)
-        if answer.slide_id:
-            result['channel_id'] = answer.slide_id.channel_id
+        if answer.slide_partner_id and answer.slide_partner_id.slide_id:
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides_survey/controllers/survey.py:10]: Is it really important to change this? Do we need the add the channel_id if you're no longer a member of this course?

File: addons/website_slides_survey/models/slide_channel.py [hunk 6]
```diff
@@ -2,9 +2,29 @@
...
+from odoo.osv import expression
 
 
 class Channel(models.Model):
     _inherit = 'slide.channel'
 
     nbr_certification = fields.Integer("Number of Certifications", compute='_compute_slides_statistics', store=True)
+
+    def _remove_membership(self, partner_ids):
+        """Unlink user_input from the slide.
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_channel.py:14]: "unlink" sounds a bit much like we are deleting things.

Would reword to something like "Remove the relationship between the user_input and the slide".
(To apply to commit message as well).

File: addons/website_slides_survey/models/slide_channel.py [hunk 7]
```diff
@@ -2,9 +2,29 @@
...
+        for channel in self:
+            removed_channel_partner_domain = expression.OR([
+                removed_channel_partner_domain,
+                [('partner_id', 'in', partner_ids),
+                 ('channel_id', '=', channel.id)]
+            ])
+        if (removed_channel_partner_domain):
+            slide_partner_ids = self.env['slide.slide.partner'].sudo().search(
+                removed_channel_partner_domain)
+            slide_partner_ids.user_input_ids.slide_id = None
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_channel.py:29]: Can't we remove the "slide_partner_id" as well? That way it becomes a regular user input no longer linked to a membership?

If we do that, we can even simplify the line below as we don't even need to filter anymore.

File: addons/website_slides_survey/tests/test_course_certification_failure.py [hunk 8]
```diff
@@ -98,11 +101,14 @@ def test_course_certification_failure_flow(self):
...
         self.assertTrue(slide_partner.survey_scoring_success, 'Quizz should be marked as passed with correct answers')
+        self.assertListEqual([1, 2, 1], slide_partner.user_input_ids.mapped(
```
Comment (awa-odoo) [inline addons/website_slides_survey/tests/test_course_certification_failure.py:110]: I would store each attempt into its own variable throughout the test and then test the attempts_number individually with a small explanation, this is currently very hard to follow.

**Review by awa-odoo — COMMENTED** [review]
Hello @Mtaylorr @flch-odoo 

I have some additional comments but I do indeed think that this is going in the right direction.

Thanks for the followup on this and cheers!

File: addons/website_slides_survey/controllers/survey.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mtaylorr) [reply addons/website_slides_survey/controllers/survey.py:10]: Indeed , this is used in the `Go Back to course` button , so when a user fails its last attempt it will be removed from the course and then this function is called so in the old behavior it would not return anything and thus the button doesn't show in the last attempt.

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides_survey/controllers/survey.py:10]: How did it work before we made the change that archive subscriptions instead of deleting them then? As the membership would not exist either at this point?

[the remark this answers is not in the document]
Comment (Mtaylorr) [reply addons/website_slides_survey/controllers/survey.py:10]: As far as I know , before the membership doesn't exist because it was archived but the answers still exist and not changed that's why it's valid to write answer.slide_id

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides_survey/controllers/survey.py:10]: Can we do that instead then, remove the membership relationship but not the slide_id, that way we have the same behavior?

[the remark this answers is not in the document]
Comment (Mtaylorr) [reply addons/website_slides_survey/controllers/survey.py:10]: Okay I will change it then

[the remark this answers is not in the document]
Comment (Mtaylorr) [reply addons/website_slides_survey/controllers/survey.py:10]: @awa-odoo  i think there is a problem with this is that removing the relationship between user_input and slide_partner_id will make it impossible to retrieve the the user_inputs for some slide_partner

File: addons/website_slides_survey/models/slide_channel.py (hunk 7 above)
Comment (flch-odoo) [reply addons/website_slides_survey/models/slide_channel.py]: I'm a bit confused about removing the `slide_partner_id` on the user input. Isn't it even more information-reducing than removing the `invite_token`? 
Didn't we want to keep the information between the attempt and a membership? The complexity added by looking one step around to get the `slide_id` looked like it was OK for the better preservation of the record's integrity and context. How will one track which are the attempts of a user? (What's even the point of keeping the invite token if you don't know who it belongs to?)

Comment (awa-odoo) [reply addons/website_slides_survey/models/slide_channel.py]: Well when we discussed this we agreed that the important part of archiving the course membership was to keep the course *progress*, not the actual certification attempts.

So I'm suggesting to get the same behavior as when we unlinked (deleted) the membership, as the code and use cases worked fine before then.
It's honestly simpler and makes sense, and the attempts will go back to being non-linked to a membership, just like it was before, no-one ever complained about that to my knowledge.

Comment (Mtaylorr) [reply addons/website_slides_survey/models/slide_channel.py]: @awa-odoo @flch-odoo  , I made the changes by removing the relationship between user_input and slide_partner_id, let me know what you think , thanks

**Comment by Mtaylorr** [comment]
Hello @flch-odoo , @awa-odoo , any news about this fix ?

File: addons/website_slides_survey/models/slide_channel.py [hunk 9]
```diff
@@ -2,9 +2,28 @@
...
+        track of the current pool of attempts allowed since the user (last) joined
+        the course, as only those will have a slide_partner_id."""
+        removed_channel_partner_domain = []
+        for channel in self:
+            removed_channel_partner_domain = expression.OR([
+                removed_channel_partner_domain,
+                [('partner_id', 'in', partner_ids), ('channel_id', '=', channel.id)]
+            ])
+        if removed_channel_partner_domain:
+            slide_partner_ids_sudo = self.env['slide.slide.partner'].sudo().search(
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_channel.py:26]: "slide_partner_ids_sudo" -> "slide_partners_sudo".

We use "ids" when it's a list of ints (or for relations names), here we have a browserecord.

File: addons/website_slides_survey/models/slide_channel.py [hunk 10]
```diff
@@ -2,9 +2,28 @@
...
+                removed_channel_partner_domain)
+            slide_partner_ids_sudo.user_input_ids.slide_partner_id = None
+        return super(Channel, self)._remove_membership(partner_ids)
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_channel.py:29]: Preferred notation is now simply "super().method()", we don't specify classname and self anymore.

File: addons/website_slides_survey/models/slide_slide.py [hunk 11]
```diff
@@ -119,8 +119,8 @@ def _generate_certification_url(self):
             if slide.channel_id.is_member:
                 user_membership_id_sudo = slide.user_membership_id.sudo()
                 # Users can Rejoin the course and he/she may have the previous Inputs
-                # Don't consider the Completed User input and Pick the only uncompleted one.
-                if user_membership_id_sudo.user_input_ids.filtered(lambda user_input: user_input.state != 'done'):
+                # The relationship between the membership and previous user_input is removed.
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_slide.py:122]: I would actually remove both comment lines, they don't make much sense anymore, we just consider existing inputs from the membership now.

File: addons/website_slides_survey/models/slide_channel.py [hunk 12]
```diff
@@ -2,9 +2,28 @@
...
+        removed_channel_partner_domain = []
+        for channel in self:
+            removed_channel_partner_domain = expression.OR([
+                removed_channel_partner_domain,
+                [('partner_id', 'in', partner_ids), ('channel_id', '=', channel.id)]
+            ])
+        if removed_channel_partner_domain:
+            slide_partner_ids_sudo = self.env['slide.slide.partner'].sudo().search(
+                removed_channel_partner_domain)
+            slide_partner_ids_sudo.user_input_ids.slide_partner_id = None
```
Comment (awa-odoo) [inline addons/website_slides_survey/models/slide_channel.py:28]: Would write "False" and not "None".

**Review by awa-odoo — COMMENTED** [review]
Hello @Mtaylorr 

Looking good, sorry for the delay.
I left a few comments to polish a bit more and then we should be good to proceed.

Thanks!

**Comment by awa-odoo** [comment]
@robodoo r+
